qualtran/resource_counting/bloq_counts.py

```diff
@@ -69,66 +69,121 @@ def _build_cbloq_counts_graph(cbloq: CompositeBloq) -> Set[BloqCountT]:
 
-    sigma: Dict[Bloq, Union[int, sympy.Expr]] = defaultdict(lambda: 0)
-    for child, n in count_decomp:
-        child = generalizer(child)
-        if child is None:
-            continue
+    # Base case 3: Empty list of callees
+    if not callee_counts:
+        raise DecomposeTypeError("Empty list of callees.")
+
+    for callee, n in callee_counts:
+        # Quite important: we do the recursive call first before adding in the edges.
+        # Otherwise, adding the edge would mark the callee node as already-visited by
+        # virtue of it being added to the graph with the `g.add_edge` call.
+
+        # Do the recursive step, which will continue to mutate `g`
+        _build_call_graph(callee, generalizer, ssa, keep, max_depth, g, depth + 1)
 
         # Update edge in `g`
-        if (parent, child) in g.edges:
-            g.edges[parent, child]['n'] += n
+        if (bloq, callee) in g.edges:
+            g.edges[bloq, callee]['n'] += n
         else:
-            g.add_edge(parent, child, n=n)
+            g.add_edge(bloq, callee, n=n)
 
-        # Do the recursive step, which will continue to mutate `g`
-        child_counts = _recurse_call_graph(child, g, ssa, generalizer, keep, max_depth, depth + 1)
 
-        # Update `sigma` with the recursion results.
-        for k in child_counts.keys():
-            sigma[k] += child_counts[k] * n
+def _build_call_graph(
+    bloq: Bloq,
+    generalizer: Callable[[Bloq], Optional[Bloq]],
+    ssa: SympySymbolAllocator,
+    keep: Callable[[Bloq], bool],
+    max_depth: Optional[int],
+    g: nx.DiGraph,
+    depth: int,
+) -> None:
+    """Recursively build the call graph.
+
+    Arguments are the same as `get_bloq_call_graph`, except `g` is the graph we're building
+    (i.e. it is mutated by this function) and `depth` is the current recursion depth.
+    """
+    if bloq in g:
+        # We already visited this node.
+        return
+
+    # Make sure this node is present in the graph. You could annotate
+    # additional node properties here, too.
+    g.add_node(bloq)
 
+    try:
+        return _build_call_graph_inner(bloq, generalizer, ssa, keep, max_depth, g, depth)
+    except (DecomposeNotImplementedError, DecomposeTypeError):
+        # We don't do anything bespoke for leaf nodes, but we could...
+        return
+
+
+def _compute_sigma(bloq, g: nx.DiGraph) -> Dict[Bloq, Union[int, sympy.Expr]]:
+    """Do a separate recursion to add up the counts of leaf bloqs.
+
+    In contrast to building the call graph, we want to visit a node as many times
+    as it has incident edges.
+
+    The fact that this is a separate step will likely be made explicit in future generalizations
+    of 'resource counting'.
+    """
+    callees = list(g.successors(bloq))
+    if not callees:
+        return {bloq: 1}
+
+    sigma: Dict[Bloq, Union[int, sympy.Expr]] = defaultdict(lambda: 0)
+    for callee in callees:
+        callee_sigma = _compute_sigma(callee, g)
```

Comment:
The current algorithm is `O(number of paths)` which can be exponential and is therefore _very_ inefficient. Note that `sigma` for each `bloq` in the same graph `g` remains the same. So we should compute it only when we visit the `bloq` for the first time and then reuse the previously computed values when we visit the `bloq` again. This will reduce the complexity to `O(n + m)` (no. of nodes + no. of edges). 


Here's an alternate iterative implementation which is fairy straightforward to generalize by generalizing (1) logic to get a default for leaves and (2) logic to combine the values of parent from value of it's children. 
```python
def _compute_sigma(bloq, g: nx.DiGraph) -> Dict[Bloq, Union[int, sympy.Expr]]:
    reverse_topo_sort = list(reversed(list(nx.topological_sort(g))))
    all_sigmas = Dict[Bloq, Dict[Bloq, Union[int, sympy.Expr]]] = defaultdict(lambda: {})
    for node in reverse_topo_sort:
        callees = list(g.successors(bloq))
        sigma = all_sigmas[node]
        if not callees:
            # 1. Default value for leaf nodes
            sigma[node] = 1
            continue
        for callee in callees:
            callee_sigma = all_sigmas[callee]
            # 2. Combination logic to compute sigma of the parent using sigma of the child.
            n = g.edges[bloq, callee]['n']
            for k in callee_sigma.keys():
                sigma[k] += callee_sigma[k] * n
    return all_sigmas[bloq]
```

Note that the recursive implementation with caching has an advantage in the scenario when you wish to compute `sigma` of a node in the graph that can reach only a small subset of nodes instead of reaching _all_ nodes of the graph. Given that for our use case, we are often interested in sigma of the root node, that's not the case and therefore the iterative implementation would almost always be faster than recursive.

Reply:
done